appcheck/appcheck.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2022 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package appcheck provides functionality for verifying App Check tokens.
+package appcheck
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/MicahParks/keyfunc"
+	"github.com/golang-jwt/jwt/v4"
+
+	"firebase.google.com/go/v4/internal"
+)
+
+// JWKSUrl is the URL of the JWKS used to verify App Check tokens.
+var JWKSUrl = "https://firebaseappcheck.googleapis.com/v1beta/jwks"
+
+const appCheckIssuer = "https://firebaseappcheck.googleapis.com/"
+
+var (
+	// ErrIncorrectAlgorithm is returned when the token is signed with a non-RSA256 algorithm.
+	ErrIncorrectAlgorithm = errors.New("token has incorrect algorithm")
+	// ErrTokenType is returned when the token is not a JWT.
+	ErrTokenType = errors.New("token has incorrect type")
+	// ErrTokenClaims is returned when the token claims cannot be decoded.
+	ErrTokenClaims = errors.New("token has incorrect claims")
+	// ErrTokenAudience is returned when the token audience does not match the current project.
+	ErrTokenAudience = errors.New("token has incorrect audience")
+	// ErrTokenIssuer is returned when the token issuer does not match Firebase's App Check service.
+	ErrTokenIssuer = errors.New("token has incorrect issuer")
+	// ErrTokenSubject is returned when the token subject is empty or missing.
+	ErrTokenSubject = errors.New("token has empty or missing subject")
+)
+
+// DecodedAppCheckToken represents a verified App Check token.
+//
+// DecodedAppCheckToken provides typed accessors to the common JWT fields such as Audience (aud) // and ExpiresAt (exp).
+// Additionally it provides an AppID field, which indicates the application ID to which this
+// token belongs. Any additional JWT claims can be accessed via the Claims map of Token.
```

Comment:
```suggestion
// token belongs. Any additional JWT claims can be accessed via the Claims map of DecodedAppCheckToken.
```

super nit. I think this was copypasta from my part.